Documentation/ceph-best-practices.md

```diff
@@ -0,0 +1,516 @@
+Provisioner is best started on the same nodes as other Ceph daemons.
+
+**`operator.yaml`:**
+```yaml
+CSI_PROVISIONER_NODE_AFFINITY: “storage-node=true”
+AGENT_NODE_AFFINITY: “storage-node=true”
+DISCOVER_AGENT_NODE_AFFINITY: “storage-node=true”
+```
+
+For `all` Ceph daemons, we must add Affinity for storage nodes in `cluster.yaml`.
+
+**`cluster.yaml`:**
+```yaml
+placement:
+  all:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+          - matchExpressions:
+              - key: storage-node
+                operator: In
+                values:
+                  - "true"
+```
+
+## Segregating Ceph from User Applications
+
+You could also have reason to totally separate Rook and Ceph nodes from application nodes. This
+falls under the disaggregated use-case, and it is a more traditional way to deploy storage. In this
+case, we still need to follow the best practices described in the section above to
+[Limit Ceph to Specific Nodes](#limiting-ceph-to-specific-nodes), and we also need some additional
+settings.
+
+To segregate Ceph from user applications, we will also label all non-storage nodes with
+`storage-node=false`. The CSI plugin pods must run where user application Pods run and not where
+Rook or Ceph Pods are run. Add a CSI plugin Affinity for all non-storage nodes in the Rook
+operator configuration.
+
+**`operator.yaml`:**
+```yaml
+CSI_PLUGIN_NODE_AFFINITY: "storage-node=false"
+```
+
+In addition to that, we will set Kubernetes node Taints and configure Rook Tolerations. For example,
+Taint the storage nodes with `storage-node=true:NoSchedule`, and then add the Tolerations shown
+below to the Rook operator in `operator.yaml`.
+
+**`operator.yaml`:**
+```yaml
+AGENT_TOLERATIONS: |
+  - key: storage-node
+    operator: Exists
+```
+```yaml
+DISCOVER_TOLERATIONS: |
+  - key: storage-node
+    operator: Exists
+```
+```yaml
+CSI_PROVISIONER_TOLERATIONS: |
+  - key: storage-node
+    operator: Exists
+```
+
+We must also add a Toleration for `all` Ceph daemon Pods in `cluster.yaml`.
+
+**`cluster.yaml`:**
+```yaml
+placement:
+  all:
+    tolerations:
+      - key: storage-node
+        operator: Exists
+```
+
+## Setting Ceph CRUSH Map via Kubernetes Node Labels
+
+A feature that was implemented early in Rook’s development is to set Ceph’s CRUSH map via Kubernetes
+Node labels. For our example data center, we recommend labelling Nodes with `room`, `rack`, and
+`chassis`.
+
+As a note, Rook will only set a CRUSH map location for each OSD upon initial creation of the OSD.
+It will not alter the CRUSH map if these labels are modified later. Therefore, modifying the CRUSH
+location of an OSD after Rook has created it must be done manually.
+
+As an example, in our hypothetical data center, labeling nodes will look like the following:
+
+```sh
+# -- room-1 --
+
+kubectl label node node-1-1-1 topology.rook.io/room=room-1
```

Comment:
Setting the topology labels on **one** of the nodes seems sufficient for the example. Showing all the nodes is very verbose.